plugin/attributes.go

```diff
@@ -0,0 +1,189 @@
+	if err != nil {
+		return nil, err
+	}
+	delete(unknownFields, constDisableCredentialRotation)
+
+	if len(unknownFields) != 0 {
+		return nil, fmt.Errorf("unknown catalog attribute fields provided: %s", keysAsString(unknownFields))
+	}
+
+	return result, nil
+}
+
+// CatalogSecrets is a Go-native representation of the Secrets map.
+type CatalogSecrets struct {
+	AccessKeyId     string
+	SecretAccessKey string
+}
+
+func getCatalogSecrets(in *structpb.Struct) (*CatalogSecrets, error) {
+	unknownFields := structFields(in)
+	result := new(CatalogSecrets)
+
+	var err error
+	result.AccessKeyId, err = getStringValue(in, constAccessKeyId, true)
+	if err != nil {
+		return nil, err
+	}
+	delete(unknownFields, constAccessKeyId)
+
+	result.SecretAccessKey, err = getStringValue(in, constSecretAccessKey, true)
+	if err != nil {
+		return nil, err
+	}
+	delete(unknownFields, constSecretAccessKey)
+
+	if len(unknownFields) != 0 {
+		return nil, fmt.Errorf("unknown catalog secret fields provided: %s", keysAsString(unknownFields))
+	}
+
+	return result, nil
+}
+
+// SetAttributes is a Go-native representation of the Attributes map that can be
+// used for decoding the incoming map via mapstructure.
+type SetAttributes struct {
+	Filters []string
+}
+
+func getSetAttributes(in *structpb.Struct) (*SetAttributes, error) {
+	var setAttrs SetAttributes
+
+	// Quick validation to ensure that there's no non-filter attributes
+	// here for now. Make this more complex if we add more attributes
+	// to host sets.
+	unknownFields := structFields(in)
+	delete(unknownFields, constDescribeInstancesFilters)
+	if len(unknownFields) != 0 {
+		return nil, fmt.Errorf("unknown set attribute fields provided: %s", keysAsString(unknownFields))
+	}
+
+	// Mapstructure complains if it expects a slice as output and sees a scalar
+	// value. Rather than use WeakDecode and risk unintended consequences, I'm
+	// manually making this change if necessary.
+	inMap := in.AsMap()
+	if filtersRaw, ok := inMap[constDescribeInstancesFilters]; ok {
+		switch filterVal := filtersRaw.(type) {
+		case string:
+			inMap[constDescribeInstancesFilters] = []string{filterVal}
+		}
+	}
+
+	if err := mapstructure.Decode(inMap, &setAttrs); err != nil {
+		return nil, fmt.Errorf("error decoding set attributes: %w", err)
+	}
+
+	return &setAttrs, nil
+}
+
+func getStringValue(in *structpb.Struct, k string, required bool) (string, error) {
+	mv := in.AsMap()
+	v, ok := mv[k]
+	if !ok {
+		if required {
+			return "", fmt.Errorf("missing required value %q", k)
+		}
+
+		return "", nil
+	}
+
+	s, ok := v.(string)
+	if !ok {
+		return "", fmt.Errorf("unexpected type for value %q: want string, got %T", k, v)
+	}
+
+	if s == "" && required {
+		return "", fmt.Errorf("value %q cannot be empty", k)
+	}
+
+	return s, nil
+}
+
+func getBoolValue(in *structpb.Struct, k string, required bool) (bool, error) {
+	mv := in.AsMap()
```

Comment:
nit (feel free to ignore): Consider not calling AsMap since that iterates over all the fields of an unvalidated user input prior to validation.  Instead rely on the native proto functions for structs and values.

```
v, ok := s.GetAttributes().GetFields()["something"]
if !ok {
    if required {
			return false, fmt.Errorf("missing required value %q", k)
		}
}    
switch v.GetKind().(type) {
case *structpb.Value_StringValue:
	v.GetStringValue()
default:
	return nil, err
}
... 
```

Reply:
Yeah, let me check, kind of wondering why I decided to do AsMap instead of looking up the value directly.

Reply:
Changed! I think the main reason was just no easy way to query a value as interface from structpb, but using GetFields/AsInterface is easy enough.